app/src/main/java/org/openobservatory/ooniprobe/model/database/Measurement.java

```diff
@@ -64,6 +68,21 @@ public Measurement(Result result, String test_name, String report_id) {
 		start_time = new java.util.Date();
 	}
 
+    public static long countNotUploaded(@Nullable Integer resultId) {
+        Where<Measurement> where = SQLite.selectCountOf().from(Measurement.class)
+                .where(Measurement_Table.is_failed.eq(false))
+                .and(Measurement_Table.is_rerun.eq(false))
+                .and(Measurement_Table.is_done.eq(true))
+                .and(OperatorGroup.clause()
+                        .or(Measurement_Table.is_uploaded.eq(false))
+                        .or(Measurement_Table.report_id.isNull())
+                );
+        if (resultId != null) {
+            where.and(Measurement_Table.result_id.eq(resultId));
+        }
```

Comment:
Please, do not do this. Passing `null` in this way is an anti pattern. I would instead recommend:

```Java
public Where<Measurement> selectUploadable() {
    return SQLite.select().from(Measurement.class)
                .where(Measurement_Table.is_failed.eq(false))
                .and(Measurement_Table.is_rerun.eq(false))
                .and(Measurement_Table.is_done.eq(true))
                .and(OperatorGroup.clause()
                        .or(Measurement_Table.is_uploaded.eq(false))
                        .or(Measurement_Table.report_id.isNull())
                );
}
```

Then, there are two options:

1. use _directly_ this method and do the right thing in place, e.g.

```Java
    @OnItemSelected(R.id.filterTests)
    void queryList() {
        if (((Application) getActivity().getApplication()).getPreferenceManager().isManualUploadResults() &&
                Measurement.selectUploadable().count()  != 0)
            snackbar.show();
        else
            snackbar.dismiss();
```

and

```Java
        adapter.notifyTypesChanged();
        if (((Application) getApplication()).getPreferenceManager().isManualUploadResults() &&
                Measurement.selectUploadable().and(Measurement_Table.result_id.eq(resultId)).count() != 0)
            snackbar.show();
        else
            snackbar.dismiss();
```

and please note that this also allows us to factor code in MKCollectorResubmitTask.java

2. perhaps write a wrapper method like this:

```Java
public Where<Measurement> selectUploadableWithResultId() {
  return Measurement.selectUploadable().and(Measurement_Table.result_id.eq(resultId));
}
```

I am undecided whether this is needed or not. I would let the code decide for us: if we do this more than once or we anticipate we would, then let's factor it, otherwise let's not bother doing that.

- - -

As for _why_ I am asking you to restructure the code as mentioned:

1. having less conditionals in our code is good because we can cover more lines with less tests and be more confident that our code works okay, because there are less corner cases to deal with

2. having conditionals based on `null`s is IMHO a signal of code stink and should be avoided, at least because when you see `method(null)` you don't know what the code is doing until you read and understand the implementation of the method and what's the impact of the `null` on that

3. writing code based on adding building blocks (i.e. a method such as the one I proposed) is good because you combine building blocks by addition, whereas having conditionals based on `null`s is more tricky because the addition is not performed only by chaining methods but there is also subtraction (which is a form of addition) that is performed as part of passing `null` arguments to specific methods